internal/subscriber/subscriber.go

```diff
@@ -12,96 +16,369 @@ import (
+	}
+
+	if shouldCommit {
+		_, err := s.consumer.CommitByPartitionAndOffset(ctx, messagebroker.CommitOnTopicRequest{
+			Topic:     req.Topic,
+			Partition: req.Partition,
+			Offset:    req.Offset,
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	removeMessageFromMemory(stats, req.MessageID)
+
+	// after successful commit to broker, make sure to re-init the maxCommittedOffset in subscriber
+	stats.maxCommittedOffset = req.Offset
+
 	return nil
 }
 
-// ModifyAckDeadline for messages
-func (c *Subscriber) ModifyAckDeadline(ctx context.Context, req *ModifyAckDeadlineRequest) error {
+// cleans up all occurrences for a given msgId from the internal data-structures
+func removeMessageFromMemory(stats *ConsumptionMetadata, msgID string) {
+	delete(stats.consumedMessages, msgID)
+
+	// remove message from offsetBasedMinHeap
+	indexOfMsgInOffsetBasedMinHeap := stats.offsetBasedMinHeap.MsgIDToIndexMapping[msgID]
+	msg := heap.Remove(&stats.offsetBasedMinHeap, indexOfMsgInOffsetBasedMinHeap).(*customheap.AckMessageWithOffset)
+	delete(stats.offsetBasedMinHeap.MsgIDToIndexMapping, msgID)
+	heap.Init(&stats.offsetBasedMinHeap)
+
+	// remove same message from deadlineBasedMinHeap
+	indexOfMsgInDeadlineBasedMinHeap := stats.deadlineBasedMinHeap.MsgIDToIndexMapping[msg.MsgID]
+	heap.Remove(&stats.deadlineBasedMinHeap, indexOfMsgInDeadlineBasedMinHeap)
+	delete(stats.deadlineBasedMinHeap.MsgIDToIndexMapping, msgID)
+	heap.Init(&stats.deadlineBasedMinHeap)
+}
+
+// modifyAckDeadline for messages
+func (s *Subscriber) modifyAckDeadline(ctx context.Context, req *ModAckMessage) error {
+
+	logger.Ctx(ctx).Infow("got mod ack request", "mod_ack_request", req.String())
+
+	tp := req.ackMessage.ToTopicPartition()
+	stats := s.consumedMessageStats[tp]
+
+	deadlineBasedHeap := stats.deadlineBasedMinHeap
+	if deadlineBasedHeap.IsEmpty() {
+		return nil
+	}
+
+	if req.ackDeadline == 0 {
+		// modAck with deadline = 0 means nack
+		// https://github.com/googleapis/google-cloud-go/blob/pubsub/v1.10.0/pubsub/iterator.go#L348
+
+		msgID := req.ackMessage.MessageID
+		msg := stats.consumedMessages[msgID].(*messagebroker.ReceivedMessage)
+
+		// push to retry queue
+		s.retry(ctx, msg.Data)
+
+		// cleanup message from memory
+		removeMessageFromMemory(stats, msgID)
+
+		return nil
+	}
+
+	indexOfMsgInDeadlineBasedMinHeap := deadlineBasedHeap.MsgIDToIndexMapping[req.ackMessage.MessageID]
+
+	// update the deadline of the identified message
+	deadlineBasedHeap.Indices[indexOfMsgInDeadlineBasedMinHeap].AckDeadline = req.ackDeadline
+	heap.Init(&deadlineBasedHeap)
+
 	return nil
 }
 
+func (s *Subscriber) checkAndEvictBasedOnAckDeadline(ctx context.Context) {
+
+	// do a deadline based eviction for all active topic-partition heaps
+	for _, metadata := range s.consumedMessageStats {
+
+		deadlineBasedHeap := metadata.deadlineBasedMinHeap
+		if deadlineBasedHeap.IsEmpty() {
+			return
+		}
+
+		// peek deadline heap
+		peek := deadlineBasedHeap.Indices[0]
```

Comment:
noticed panic here. You might want to check length of the array here.

Reply:
Already handled at : https://github.com/razorpay/metro/pull/64/files#diff-4bcad8e79804b094286e6698fe58e6f08d402b61a217a004fd4b078034ee9992R209
We deployed an older commit.